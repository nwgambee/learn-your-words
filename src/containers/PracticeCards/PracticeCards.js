import React, { Component } from 'react';
import { connect } from 'react-redux';
import './PracticeCards.scss'
import PropTypes from 'prop-types';

export class PracticeCards extends Component {
  constructor() {
    super();
    this.state = {guess: '', correctlyGuessedWords: []}
 }

 handleChange = e => {
   this.setState({ [e.target.name]: e.target.value })
 }

 shuffleList(array) {
  var currentIndex = array.length, temporaryValue, randomIndex;

  while (0 !== currentIndex) {

    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;

    temporaryValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = temporaryValue;
  }
  return array;
}
checkGuess = (word, e) => {
  if(this.state.guess.toLowerCase() === word.toLowerCase()) {
    this.setState({ guess: '', correctlyGuessedWords: [...this.state.correctlyGuessedWords, word.toLowerCase()]})

  } else {
    console.log('incorrect guess');
    // prompt 'incorrect guess' message
  }
}

 render() {
    let shuffledList = this.shuffleList(this.props.vocabList);
    let cards = shuffledList.map(word => {
      if(this.state.correctlyGuessedWords.includes(word.word.toLowerCase())) {
        return (
          <section className='list-card' key={word.word}>
            <h1 className='list-word-input'> {word.word}</h1>
            <h2 className='list-definition'>{word.results[0].definition}</h2>
            <h2 className='list-part-of'>{word.results[0].partOfSpeech}</h2>
          </section> 
        )  
      } 
     return (
            <section className='list-card' key={word.word}>
              <input className='list-word-input'
               type='text'
               placeholder='Word'
               name={`guess`}
               onChange={this.handleChange}
              />
              <button className='check-guess' onClick={(e) => this.checkGuess(word.word, e)}>Check Guess</button>
              <h2 className='list-definition'>{word.results[0].definition}</h2>
              <h2 className='list-part-of'>{word.results[0].partOfSpeech}</h2>
            </section> 
 
    )
  })

   return (
     <section className='list-container'>
       {cards}
     </section>
   )
 }
} 

export const mapStateToProps = state => ({
  vocabList: state.vocabList
})

export default connect(mapStateToProps)(PracticeCards)

PracticeCards.propTypes = {
  vocabList: PropTypes.array
}