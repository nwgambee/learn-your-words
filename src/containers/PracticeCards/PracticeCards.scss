.list-word-input {
  text-align: center;
  font-size: 4rem;
  margin: 0;
  border: 2px solid black;
  border-radius: .4rem;
  padding-bottom: .3rem;
}

.check-guess {
  margin-top: 1rem;
  background-color: #35D461;
  padding: .75rem;
  font-size: 1.5rem;
  width: 20rem;
  border: 5px solid #F99D07 ;
  border-radius: .5rem;
  cursor: pointer;
  font-weight: 700;
  align-self: center;
}

.correct-message {
  color: #F99D07;
  font-weight: 900;
  font-size: 2rem;
}
