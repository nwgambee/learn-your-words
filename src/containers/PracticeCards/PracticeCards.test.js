import React from 'react';
import { PracticeCards, mapStateToProps, mapDispatchToProps } from './PracticeCards';
import { shallow } from 'enzyme';

describe('PracticeCards', () => {
  let wrapper;
  it('should match the snapshot', () => {
    wrapper = shallow(<PracticeCards vocabList={[{word: 'word', results: [{definition: 'definition', partOfSpeech: 'noun'}, {}]}]}/>)
    expect(wrapper).toMatchSnapshot();
  })

  describe('Method Tests', () => {
    it('should shuffle the given array when shuffleList is called', () => {
      wrapper = shallow(<PracticeCards vocabList={[{word: 'word', results: [{definition: 'definition', partOfSpeech: 'noun'}, {}]}]}/>)
      let unshuffled = [1,2,3,4]
      expect(wrapper.instance().shuffleList(unshuffled).length).toEqual(unshuffled.length)
    })
    it('should call checkGuess when the checkGuess button is clicked', () => {
      wrapper = shallow(<PracticeCards vocabList={[{word: 'word', results: [{definition: 'definition', partOfSpeech: 'noun'}, {}]}]}/>)
      wrapper.instance().checkGuess = jest.fn();
      wrapper.find('.check-guess').simulate('click');
      expect(wrapper.instance().checkGuess).toHaveBeenCalled()
    })
    it('should update state when a word is guessed correctly', () => {
      wrapper = shallow(<PracticeCards vocabList={[{word: 'word', results: [{definition: 'definition', partOfSpeech: 'noun'}, {}]}]}/>)
      const startState = {guess: 'hat', correctlyGuessedWords: []};
      wrapper.setState(startState);
      const expected = {guess: '', correctlyGuessedWords: ['hat']};
      wrapper.instance().checkGuess('hat');
      expect(wrapper.state()).toEqual(expected);
    })
    it('should update state when handleChange is called', () => {
      const startState = {guess: '', correctlyGuessedWords: []};
      wrapper = shallow(<PracticeCards vocabList={[{word: 'word', results: [{definition: 'definition', partOfSpeech: 'noun'}, {}]}]}/>)
      wrapper.setState(startState);
      const mockEvent = {target: {name: 'guess', value: 'hat'}};
      const expected = {guess: 'hat', correctlyGuessedWords: []};
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('guess')).toEqual(expected.guess);
    })
  })

  describe('mapStateToProps', () => {
    it('should return an object with the vocabList array', () => {
      const mockState = {
        searchedWord: {word: 'hat', definition: 'this is a definition', type: 'noun'},
        loadingStatus: true,
        vocabList: [{word: 'hat'}, {word: 'cat'}],
        someOtherProp: false
      }
      const expected = {
        vocabList: [{word: 'hat'}, {word: 'cat'}],
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })
})