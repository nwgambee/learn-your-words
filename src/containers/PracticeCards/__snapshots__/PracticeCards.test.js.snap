// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PracticeCards should match the snapshot 1`] = `
<section
  className="list-container"
>
  <section
    className="list-card"
    key="word"
  >
    <h1
      className="list-word-h1"
    >
      word
    </h1>
    <h2
      className="list-definition"
    >
      definition
    </h2>
    <h2
      className="list-part-of"
    >
      noun
    </h2>
    <button
      className="remove-from-list-btn"
      onClick={[Function]}
    >
      Remove From List
    </button>
  </section>
</section>
`;
